rust/src/ldap/types.rs

```diff
@@ -0,0 +1,645 @@
+                Self::from_search_result_done(msg)
+            }
+            ldap_parser::ldap::ProtocolOp::SearchResultReference(msg) => {
+                Self::from_search_result_reference(msg)
+            }
+            ldap_parser::ldap::ProtocolOp::ModifyRequest(msg) => Self::from_modify_request(msg),
+            ldap_parser::ldap::ProtocolOp::ModifyResponse(msg) => Self::from_modify_response(msg),
+            ldap_parser::ldap::ProtocolOp::AddRequest(msg) => Self::from_add_request(msg),
+            ldap_parser::ldap::ProtocolOp::AddResponse(msg) => Self::from_add_response(msg),
+            ldap_parser::ldap::ProtocolOp::DelRequest(msg) => Self::from_del_request(msg),
+            ldap_parser::ldap::ProtocolOp::DelResponse(msg) => Self::from_del_response(msg),
+            ldap_parser::ldap::ProtocolOp::ModDnRequest(msg) => Self::from_mod_dn_request(msg),
+            ldap_parser::ldap::ProtocolOp::ModDnResponse(msg) => Self::from_mod_dn_response(msg),
+            ldap_parser::ldap::ProtocolOp::CompareRequest(msg) => Self::from_compare_request(msg),
+            ldap_parser::ldap::ProtocolOp::CompareResponse(msg) => Self::from_compare_response(msg),
+            ldap_parser::ldap::ProtocolOp::ExtendedRequest(msg) => Self::from_extended_request(msg),
+            ldap_parser::ldap::ProtocolOp::ExtendedResponse(msg) => {
+                Self::from_extended_response(msg)
+            }
+            ldap_parser::ldap::ProtocolOp::IntermediateResponse(msg) => {
+                Self::from_intermediate_response(msg)
+            }
+            _ => ProtocolOp::Unknown,
+        };
+        let controls = ldap_msg.controls.map(|ctls| {
+            ctls.iter()
+                .map(|ctl| Control {
+                    control_type: LdapOID(ctl.control_type.0.to_string()),
+                    criticality: ctl.criticality,
+                    control_value: ctl.control_value.as_ref().map(|val| val.to_vec()),
+                })
+                .collect()
+        });
+
+        Self {
+            message_id,
+            protocol_op,
+            controls,
+        }
+    }
+}
+
+impl LdapMessage {
+    pub fn is_unknown(&self) -> bool {
+        match self.protocol_op {
+            ProtocolOp::Unknown => return true,
+            _ => return false,
+        }
+    }
+
+    pub fn is_request(&self) -> bool {
+        match self.protocol_op {
+            ProtocolOp::BindRequest(_)
+            | ProtocolOp::UnbindRequest
+            | ProtocolOp::SearchRequest(_)
+            | ProtocolOp::ModifyRequest(_)
+            | ProtocolOp::AddRequest(_)
+            | ProtocolOp::DelRequest(_)
+            | ProtocolOp::ModDnRequest(_)
+            | ProtocolOp::CompareRequest(_)
+            | ProtocolOp::ExtendedRequest(_) => {
+                return true;
+            }
+            _ => {
+                return false;
+            }
+        }
+    }
+
+    pub fn is_response(&self) -> bool {
```

Comment:
Is this not the opposite of is_request ?

Reply:
Maybe I will make it explicit in a follow-up commit as a nit improvement, by returning `!self.is_request()` instead of listing the types again